### ISSUE TITLE
Exclusion failures in Sphinx 4

### ISSUE THREAD

**Issue by dkfellows** [sphinx-doc/sphinx#9211]
**Describe the bug**
Sphinx version 4.0.0 (or perhaps apidoc?) is causing an additional warning that was not present in previous versions of Sphinx (3.5). This is an issue when using Sphinx in CI workflows to validate the docs, as those are set to report warnings as errors.

The core of the issue (described further below) is that I can't figure out _why_ Sphinx 4 +autodoc has decided that it needs to have a reference to `foo.bar.grill.Grill`; I _specifically_ want that documented as `foo.bar.Grill`. (It's in a separate file simply to enable management of project complexity; the real code that this is cut down from has much deeper package paths and far far larger classes).

----

**To Reproduce**

```
git clone https://github.com/dkfellows/sphinx4bug
cd sphinx4bug
pip install -r requirements.txt
make clean html
```

Note that a warning is generated. Here's the version from on my system:

`/Users/dkf/git/sphinx4bug/foo/example.py:docstring of foo.example.FooBarExample.make_grill:: WARNING: more than one target found for cross-reference 'Grill': foo.bar.Grill, foo.bar.grill.Grill`

Exact path will vary, of course.

Full log from my system:
```
dkf$ make clean html
rm -f -rf foo.rst foo.bar.rst modules.rst _build 
Running Sphinx v4.0.0
FILTERING FILE IN __init__: foo/bar/grill.py
making output directory... done
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 4 source files that are out of date
updating environment: [new config] 4 added, 0 changed, 0 removed
reading sources... [100%] modules                                               
looking for now-outdated files... none found
pickling environment... done
checking consistency... done
preparing documents... done
writing output... [100%] modules                                                
/Users/dkf/git/sphinx4bug/foo/example.py:docstring of foo.example.FooBarExample.make_grill:: WARNING: more than one target found for cross-reference 'Grill': foo.bar.Grill, foo.bar.grill.Grill
generating indices... genindex py-modindex done
writing additional pages... search done
copying static files... done
copying extra files... done
dumping search index in English (code: en)... done
dumping object inventory... done
build succeeded, 1 warning.

The HTML pages are in _build/html.
```

Setting `_shorten_class_names` to `True` in `conf.py` enables a workaround (that we rather like) in the majority of cases, but it still fails weirdly in a few (such as with a documented enumeration member of a class) where I can't seem to find out _why_.

**Expected behavior**
I expected the fact that `foo/bar/grill.py` was excluded to mean that `foo.bar.grill.Grill` would not form part of the resulting documentation tree.

**Your project**
Problem manifests in https://github.com/dkfellows/sphinx4bug but is appearing for real in several repositories in https://github.com/SpiNNakerManchester

**Environment info**
- OS: macOS (probably many others; doesn't seem to be an OS-related issue)
- Python version: tested with 3.6.13 (but doesn't seem to be a Python-version related issue)
- Sphinx version: 4.0.0
- Sphinx extensions: `sphinx.ext.autodoc` (critical)
- Extra tools: none

**Additional context**

- https://github.com/SpiNNakerManchester/SpiNNFrontEndCommon/issues/805

**Comment by jmckenna** [comment]
Not sure if this is related, but Sphinx 4.0.1 was released yesterday.

**Comment by tk0miya** [comment]
Yes, this would be fixed at 4.0.1. Could you try it please?
refs: https://github.com/sphinx-doc/sphinx/issues/9205

**Comment by dkfellows** [comment]
I can confirm that this fixes the test case, and with some other changes (not the subject of this bug report) resolves the problems observed in the real repositories. It doesn't seem to fix things entirely, but it's possible that the remaining problems were real bugs in the source repo that Sphinx wasn't picking up previously. I can live with that; the changes we've applied to reduce the problem to nothingness have had other good side effects too.

The changes in question are approximately equivalent to setting `_shorten_class_names` to `True` in the sample repo and moving some classes around to what are IMHO better places anyway.

Thanks!
